### PR TITLE
index, restic: Genericize indexMap

### PR DESCRIPTION
**Pull request by greatroar**
What does this PR change? What problem does it solve?
-----------------------------------------------------

This converts the indexMap from internal/index to a generic type called restic.Map, keyed on a restic.BlobHandle but with a generic value type. It should be possible to reimplement restic.CountedBlobSet using this type as well. In fact, I already did, but I did not test or benchmark yet and the prune changes are not in this draft PR yet.

The size of a MapEntry[index.mapData] is still 64, same as the old indexEntry. So far, index lookups seem to have become slower. I've still to find out why.

<details><summary>Benchmark results for internal/index</summary><pre>
name                                             old time/op    new time/op    delta
DecodeIndex-8                                       4.13s ± 1%     4.05s ± 1%   -2.04%  (p=0.000 n=9+9)
DecodeIndexParallel-8                               4.14s ± 1%     4.07s ± 0%   -1.79%  (p=0.000 n=9+9)
IndexHasUnknown-8                                  40.5ns ±11%    50.0ns ±29%  +23.33%  (p=0.000 n=10+10)
IndexHasKnown-8                                    45.0ns ±30%    48.5ns ± 9%     ~     (p=0.093 n=10+10)
IndexAlloc-8                                        588ms ± 1%     588ms ± 1%     ~     (p=0.666 n=9+9)
IndexAllocParallel-8                                269ms ± 2%     269ms ± 1%     ~     (p=0.888 n=9+8)
MasterIndexAlloc-8                                  268ms ± 1%     269ms ± 1%   +0.58%  (p=0.008 n=10+9)
MasterIndexLookupSingleIndex-8                      137ns ± 5%     134ns ± 9%     ~     (p=0.592 n=10+10)
MasterIndexLookupMultipleIndex-8                    144ns ±12%     147ns ± 6%     ~     (p=0.271 n=10+10)
MasterIndexLookupSingleIndexUnknown-8              73.3ns ± 7%    73.7ns ±11%     ~     (p=1.000 n=10+10)
MasterIndexLookupMultipleIndexUnknown-8            69.4ns ± 9%    67.7ns ±10%     ~     (p=0.481 n=10+10)
MasterIndexLookupParallel/known,indices=25-8        499ns ± 7%     499ns ± 2%     ~     (p=0.971 n=10+10)
MasterIndexLookupParallel/unknown,indices=25-8      306ns ± 2%     388ns ± 3%  +26.87%  (p=0.000 n=10+10)
MasterIndexLookupParallel/known,indices=50-8        476ns ± 3%     491ns ± 1%   +3.18%  (p=0.000 n=10+9)
MasterIndexLookupParallel/unknown,indices=50-8      320ns ± 7%     386ns ± 3%  +20.44%  (p=0.000 n=10+10)
MasterIndexLookupParallel/known,indices=100-8       504ns ± 2%     499ns ± 3%     ~     (p=0.093 n=10+10)
MasterIndexLookupParallel/unknown,indices=100-8     326ns ± 7%     398ns ± 7%  +22.34%  (p=0.000 n=10+10)
MasterIndexLookupBlobSize-8                        59.1ns ± 7%    65.7ns ±13%  +11.05%  (p=0.000 n=10+10)
MasterIndexEach-8                                   948ms ± 1%     943ms ± 2%     ~     (p=0.436 n=9+9)

name                                             old alloc/op   new alloc/op   delta
IndexAlloc-8                                        451MB ± 0%     451MB ± 0%   -0.00%  (p=0.000 n=10+10)
IndexAllocParallel-8                                451MB ± 0%     451MB ± 0%   -0.00%  (p=0.009 n=10+9)
MasterIndexAlloc-8                                  162MB ± 0%     159MB ± 0%   -1.97%  (p=0.000 n=7+10)

name                                             old allocs/op  new allocs/op  delta
IndexAlloc-8                                        1.53M ± 0%     1.53M ± 0%     ~     (all equal)
IndexAllocParallel-8                                1.53M ± 0%     1.53M ± 0%     ~     (p=1.210 n=10+9)
MasterIndexAlloc-8                                   585k ± 0%      585k ± 0%     ~     (all equal)
</pre></details>

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Prune memory issues have been discussed several times, including at #3844.

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [ ] I'm done! This pull request is ready for review.

### CODE REVIEW

**Comment by MichaelEischer** [comment]
> It should be possible to reimplement restic.CountedBlobSet using this type as well.

I like the idea in general, although I wonder how it would be possible to embed the counter byte directly into the normal repository index. That way we could minimize the memory overhead of the BlobSet to just a few percent of the repository index.

The index could the return a CountedBlobSet (only once, to prevent multiple users from conflicting) which can store that information directly in the index.

**Comment by greatroar** [comment]
I'm a bit closer to finding out what's going on with the performance of the "unknown" lookups: it's not the generics, because when I apply the same refactor without generics to indexMap, I get the same slowdown.

File: internal/restic/map.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+		if e.ID != h.ID {
+			continue
+		}
+		fn(e)
+	}
+}
+
+// Get returns the first entry for the given id.
+func (m *Map[V]) Get(h BlobHandle) *MapEntry[V] {
+	tm := m.byType[h.Type]
```
Comment (MichaelEischer) [inline internal/restic/map.go:99]: That creates a copy of the map on the stack. Not sure though whether that is the cause of the performance problems.

Comment (greatroar) [reply internal/restic/map.go:99]: Fixed. This is a part of it, but it's not the whole story.

```
name                                             old time/op  new time/op  delta
IndexHasUnknown-8                                47.4ns ±16%  44.4ns ±11%     ~     (p=0.371 n=10+5)
MasterIndexLookupSingleIndexUnknown-8            70.1ns ± 9%  76.8ns ±11%     ~     (p=0.165 n=10+5)
MasterIndexLookupMultipleIndexUnknown-8          70.8ns ± 8%  70.4ns ± 7%     ~     (p=0.768 n=10+5)
MasterIndexLookupParallel/unknown,indices=25-8    382ns ± 2%   321ns ± 2%  -15.99%  (p=0.001 n=10+5)
MasterIndexLookupParallel/unknown,indices=50-8    369ns ± 8%   359ns ± 3%     ~     (p=0.165 n=10+5)
MasterIndexLookupParallel/unknown,indices=100-8   392ns ± 3%   364ns ± 0%   -7.08%  (p=0.002 n=10+4)
MasterIndexLookupBlobSize-8                      63.0ns ± 9%  64.3ns ± 9%     ~     (p=0.594 n=10+5)
```

**Comment by MichaelEischer** [comment]
I guess the last commit fixes the performance problems? I'm not entirely happy with adding more code to `internal/restic`, but it's probably the best place for now.

**Comment by MichaelEischer** [comment]
My current impression is that https://github.com/restic/restic/pull/4354 is more likely a better solution than a generic indexMap. In particular the `AssociatedData` type in #4354 only requires about 1/70 the memory which would be necessary when using a generic indexMap.

**Comment by MichaelEischer** [comment]
After thinking a bit more about this PR, I still think that #4354 is the way forward. Thus, I'll close the PR for now.

**Pull request closed without merging**